cassandra/pool.py

```diff
@@ -450,7 +450,19 @@ def _get_connection_for_routing_key(self, routing_key=None):
         shard_id = None
         if not self._session.cluster.shard_aware_options.disable and self.host.sharding_info and routing_key:
             t = self._session.cluster.metadata.token_map.token_class.from_key(routing_key)
-            shard_id = self.host.sharding_info.shard_id_from_token(t.value)
+            
+            shard_id = None
+            if self._session.cluster.check_tablets_enabled() and table is not None:
+                if keyspace is None:
+                    keyspace = self._keyspace
+
+                tablet = self._session.cluster._load_balancing_policy._cluster_metadata._tablets.get_tablet_for_key(keyspace, table, t)
+
+                if tablet is not None:
+                    shard_id = tablet.replicas[0][1]
```

Comment:
Why do you always take the shard of the first replica? This pool may be for a different replica.

Reply:
But how to know which shard to take? Isn't it redundant? I thought the choice didn't matter so I took the first one, maybe I misunderstood something

Reply:
Due to the current two-pass implementation (first we determine the node to send the request to, then we determine the shard to send the request to), I think this code has to go through all replicas, find the replica that matches the `HostConnection`'s `host` and read its shard number.

Reply:
@sylwiaszunejko IIUC this function chooses shard-connection to a particular host for a given key. The owning shard of a given key is different on different hosts. It's specified in the <host, shard> tuple in the replica list. So the shard should depend on which host-replica you're going to talk to.

Reply:
Ok, now I get it, hope now it looks good